pcap-rewrite/src/lib.rs

```diff
@@ -80,19 +85,60 @@ fn get_reader(input_filename: &str) -> io::Result<Box<dyn Read>> {
         Box::new(io::stdin())
     } else {
         let path = Path::new(&input_filename);
-        let file = File::open(path).map_err(|e| {
+        let mut file = File::open(path).map_err(|e| {
             error!("Could not open input file '{}'", input_filename);
             e
         })?;
-        if input_filename.ends_with(".gz") {
-            Box::new(GzDecoder::new(file))
-        } else if input_filename.ends_with(".xz") {
-            Box::new(XzDecoder::new(file))
-        } else if input_filename.ends_with(".lz4") {
-            Box::new(lz4::Decoder::new(file)?)
-        } else {
-            Box::new(file) as Box<dyn io::Read>
+
+        // https://en.wikipedia.org/wiki/LZ4_(compression_algorithm)
+        fn lz4_matcher(buf: &[u8]) -> bool {
+            buf.len() >= 4 && buf[0] == 0x04 && buf[1] == 0x22 && buf[2] == 0x4d && buf[3] == 0x18
+        }
+        // https://www.tcpdump.org/manpages/pcap-savefile.5.html
+        fn pcap_same_endianess_matcher(buf: &[u8]) -> bool {
+            buf.len() >= 4 && buf[0] == 0xa1 && buf[1] == 0xb2 && buf[2] == 0xc3 && buf[3] == 0xd4
         }
+        fn pcap_reverse_endianess_matcher(buf: &[u8]) -> bool {
+            buf.len() >= 4 && buf[0] == 0xd4 && buf[1] == 0xc3 && buf[2] == 0xb2 && buf[3] == 0xa1
+        }
+        let mut info = infer::Infer::new();
+        info.add("custom/lz4", "lz4", lz4_matcher);
+        info.add("custom/pcap", "pcap", pcap_same_endianess_matcher);
+        info.add("custom/pcap", "pcap", pcap_reverse_endianess_matcher);
+
+        let mut buf = vec![0; 4];
+        file.read_exact(&mut buf)?;
+
+        let kind = info.get(&buf).unwrap();
+
+        file.seek(SeekFrom::Start(0))?;
+        let b: Result<Box<dyn Read>, Error> = match kind.mime_type() {
+            "application/gz" => {
+                if !input_filename.ends_with(".gz") {
+                    warn!("Inferred file type is gz but file extension is not gz")
+                }
```

Comment:
Are these warnings useful? The new code would not rely on the extension at all
Unless you just want to inform the user that the extension is wrong - that could be a valid reason

Reply:
I indeed wanted to warn the user about the extension (cf personal experience :)).
Should I use info!() instead?

Reply:
If this is intentional, I'm ok with it :)